docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,230 @@
+* Extension controllers implement a mechanism which allows them to store their state or to be restored from an already existing state on a different seed cluster.
+* The already existing shoot reconciliation flow is reused for migration with minimum changes
+
+
+## Terminology
+
+__Source Seed__ is the seed which currently hosts the control plane of a Shoot Cluster
+
+__Destination Seed__ is the seed to which the control plane is being migrated
+
+
+## Resources and controller state which have to be migrated between two seeds:
+
+**Note:** The following lists are just FYI and are meant to show the current resources which need to be moved to the __Destination Seed__
+
+### Secrets
+
+Gardener has preconfigured lists of needed secrets which are generated when a shoot is created and deployed in the seed. Following is a minimum set of secrets which must be migrated to the __Destination Seed__. Other secrets can be regenerated from them.
+
+- ca
+- ca-front-proxy
+- static-token
+- ca-kubelet
+- ca-metrics-server
+- etcd-encryption-secret
+- kube-aggregator
+- kube-apiserver-basic-auth
+- kube-apiserver
+- service-account-key
+- ssh-keypair
+
+### Custom Resources and state of extension controllers
+
+Gardenlet deploys custom resources in the __Source Seed__ cluster during shoot reconciliation which are reconciled by extension controllers. The state of these controllers and any additional resources they create is independent of the gardenlet and must also be migrated to the __Destination Seed__. Following is a list of custom resources, and the state which is generated by them that has to be migrated.
+
+- **BackupBucket**: nothing relevant for migration
+- **BackupEntry**: nothing relevant for migration
+- **ControlPlane**: nothing relevant for migration
+- **DNSProvider**/DNSEntry: nothing relevant for migration
+- **Extensions**: migration of state needs to be handled individually
+- **Infrastructure**: terraform state
+- **Network**: nothing relevant for migration
+- **OperatingSystemConfig**: nothing relevant for migration
+- **Worker**: Machine-Controller-Manager related objects: machineclasses, machinedeployments, machinesets, machines
+
+This list depends on the currently installed extensions and can change in the future
+
+## Proposal
+
+### Custom Resource on the garden cluster
+
+The Garden cluster has a new Custom Resource which is stored in the project namespace of the Shoot. It contains all the required data described above so that the control plane can be recreated on the __Destination Seed__.
+This data is separated into two sections. The first is generated by the gardenlet and then either used to generate new resources (e.g secrets) or is directly deployed to the Shoot's control plane on the __Destination Seed__.
+
+The second is generated by the extension controllers in the seed.
+
+```
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: my-shoot
```

Comment:
> Will shoot have reference to this or will it have same name as the shoot? 

Same name should be good enough as there is a 1:1 mapping, and we don't want to expose it to end-users anyways. `Shoot` does also not have a reference to the `BackupEntry` we create internally.

> But just to be clear on this, its worth mentioning in doc, in shoot deletion flow, the last step should be delete this object before removing the gardener finalizer. 

If the `ShootState/ShootSnapshot` object has an owner reference pointing to the `Shoot` then we don't need to implement explicit deletion (we do the same for the shoot secrets we create in the project namespace). We only have to remove the finalizer if no `Shoot` with the same name does yet exist.

Reply:
>Same name should be good enough as there is a 1:1 mapping, and we don't want to expose it to end-users anyways. Shoot does also not have a reference to the BackupEntry we create internally.

Agreed. i also prefer the same.

> If the ShootState/ShootSnapshot object has an owner reference pointing to the Shoot then we don't need to implement explicit deletion (we do the same for the shoot secrets we create in the project namespace).

But unlike backupentry, here ShootState/ShootSnapshot should block the owner shoot deletion. 

> We only have to remove the finalizer if no Shoot with the same name does yet exist.

If i understand this statement correctly, it means there will be a situation where ShootSnapshot exists but Shoot doesn't. IMO this should never occur. Cause if shoot is deleted and ShootSnaphost is still there and Then new shoot is created with same name it will have conflicts with ShootSnapshot. Since both have same name.

Reply:
If we do not block the `Shoot` deletion and the `Shoot` gets deleted before the `ShootState/ShootSnapshot` gets deleted then the deletion logic for the `ShootState/ShootSnapshot` should just remove the finalizer.

If we only use OwnerReferences the `ShootState/ShootSnapshot` resource could be orphaned if the delete command is not cascading (e.g. `kubectl delete ... --cascade=false`). This is also true for the secrets in the project namespace. If a shoot with the same name is created in the project, then the secrets would simply be overwritten by `SyncShootCredentialsToGarden`. Same could be done by the gardenlet for the ShootState/ShootSnapshot. This won't be a problem as during creation of a new shoot migration is not triggered, therefore the restore annotation is not specified and the contents of ShootState/ShootSnapshot are completely ignored.

Reply:
I think I mentioned it already in some of our conversations, just documenting it here again: I don't understand all of this complexity. Like the kubeconfig secret for shoots that we copy into the project NS in the garden cluster, we can simply set an owner reference in the `ShootState` pointing to the `Shoot`.
Now, when deleting a shoot, we don't do anything other than releasing the finalizer from the `ShootState` object (as very last step in the flow). Now, the garbage collector cleans the "orphaned" `ShootState` object and we're done.